src/main/java/com/amazon/ion/impl/IonReaderContinuableCoreBinary.java

```diff
@@ -1369,10 +1378,301 @@ private enum State {
+                        break;
+                    case INT:
+                        switch (getIntegerSize()) {
+                            case INT:
+                            case LONG:
+                                expression = new Expression.LongIntValue(annotations, longValue());
+                                break;
+                            case BIG_INTEGER:
+                                expression = new Expression.BigIntValue(annotations, bigIntegerValue());
+                                break;
+                            default:
+                                throw new IllegalStateException();
+                        }
+                        break;
+                    case FLOAT:
+                        expression = new Expression.FloatValue(annotations, doubleValue());
+                        break;
+                    case DECIMAL:
+                        expression = new Expression.DecimalValue(annotations, decimalValue());
+                        break;
+                    case TIMESTAMP:
+                        expression = new Expression.TimestampValue(annotations, timestampValue());
+                        break;
+                    case SYMBOL:
+                        expression = new Expression.SymbolValue(annotations, symbolValue());
+                        break;
+                    case STRING:
+                        expression = new Expression.StringValue(annotations, stringValue());
+                        break;
+                    case CLOB:
+                        expression = new Expression.ClobValue(annotations, newBytes());
+                        break;
+                    case BLOB:
+                        expression = new Expression.BlobValue(annotations, newBytes());
+                        break;
+                    default:
+                        throw new IllegalStateException();
+                }
+            }
+            expressions.add(expression);
+        }
+
+        /**
+         * Reads a container value from the stream into expressions.
+         * @param type the type of container.
+         * @param annotations any annotations on the container.
+         * @param expressions receives the expressions as they are materialized.
+         */
+        private void readContainerValueAsExpression(
+            IonType type,
+            List<SymbolToken> annotations,
+            List<Expression.EExpressionBodyExpression> expressions
+        ) {
+            int startIndex = expressions.size();
+            expressions.add(Expression.Placeholder.INSTANCE);
+            stepIntoContainer();
+            while (nextValue() != Event.END_CONTAINER) {
+                if (type == IonType.STRUCT) {
+                    expressions.add(new Expression.FieldName(getFieldNameSymbol()));
+                }
+                readValueAsExpression(expressions); // TODO avoid recursion
+            }
```

Comment:
I can follow this since this method closely resembles its Rust counterpart, but I think it could use some explanatory comments for readers new to the space. It may not be obvious that the `List<Expression.EExpressionBodyExpression>` is a 'compiled' form of the e-expression for the evaluator to read.

Reply:
Added some comments.